src/utils/db_manager.rs

```diff
@@ -24,47 +37,267 @@ impl DbManager {
+                        LIMIT {limit}",
+                    ids.join(",")
+                );
+
+                // Scope needed to drop borrowing of transaction
+                {
+                    let mut stmt = transaction.prepare(&query)?;
+                    let rows = stmt.query_map(params![query_embedding.as_bytes()], |row| {
+                        row.get::<_, String>(0)
+                    })?;
+
+                    for content in rows {
+                        memories.push(content?);
+                    }
+                }
+
+                transaction.commit()?;
+
+                Ok(memories)
+            })
+            .await
+            .map_err(|e| eyre::eyre!("Failed to retrieve memories: {}", e))
+    }
+
+    // Sanity check: sqlite vec extension is loaded
+    pub async fn check_loaded_extension(database: &Connection) -> eyre::Result<()> {
+        let version: String = database
+            .call(|db| {
+                db.query_row("SELECT vec_version();", [], |row| row.get(0))
+                    .map_err(|e| e.into())
+            })
+            .await
+            .expect("Sqlite-vec extension not loaded");
+
         Ok(())
     }
 }
+
+// To do: Create in memory DB for tests purposes
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[tokio::test]
+    async fn test_db_operations() {
+        let config = Config::from_toml("./src/utils/tests/config.tests.toml")
+            .expect("Failed to load test config");
+        let database = DbManager::init_db(&config)
+            .await
+            .expect("Failed to initialize database");
+
+        let float_vec: Vec<f32> = vec![0.1, 0.2, 0.3, 0.4];
+        let text: String = "Text sample".to_string();
+        let storage_keys: HashMap<String, String> = [
+            ("key1".to_string(), "1".to_string()),
+            ("key2".to_string(), "2".to_string()),
+        ]
+        .into_iter()
+        .cloned()
+        .collect();
+
+        DbManager::insert_embedding_and_text(
+            &database,
+            text.clone(),
+            float_vec.clone(),
+            storage_keys,
+        )
+        .await
+        .expect("Failed to insert embedding and text");
+
+        let vec: Vec<f32> = vec![0.1, 0.1, 0.1, 0.1];
+        let retrieval_keys: HashMap<String, String> = [("key1".to_string(), "1".to_string())]
+            .into_iter()
+            .cloned()
+            .collect();
+        let retrieved_memories = DbManager::retrieve_memories(
+            &database,
+            vec,
+            retrieval_keys,
+            config.haiku.db_config.memory_retrieval_limit,
+        )
+        .await
+        .expect("Failed to retrieve memories");
+
+        assert_eq!(
+            retrieved_memories[0],
+            "Text sample".to_string(),
+            "No memories retrieved"
+        );
+    }
+
+    #[tokio::test]
+    async fn test_embedding_precision() {
+        let config = Config::from_toml("./src/utils/tests/config.tests.toml")
```

Comment:
let's derive Default for the `Config` type and create a default config then manually fill out needed types to test the DB

Reply:
And remove config.tests.toml

Reply:
Good idea, please check the implementation